sdb/commands/linux/internal/slub_helpers.py

```diff
@@ -122,3 +124,173 @@ def util(cache: drgn.Object) -> int:
     if total_mem == 0:
         return 0
     return int((active_memory(cache) / total_mem) * 100)
+
+
+def cache_red_left_padding(cache: drgn.Object) -> int:
+    SLAB_RED_ZONE_FLAG = 0x00000400
+    padding = 0
+    if cache.flags & SLAB_RED_ZONE_FLAG:
+        padding += int(cache.red_left_pad)
+    return padding
+
+
+def cache_get_free_pointer(cache: drgn.Object, p: drgn.Object) -> drgn.Object:
+    """
+    Get the next pointer in the freelist. Note, that this
+    function assumes that CONFIG_SLAB_FREELIST_HARDENED
+    is set in the target
+    """
+    assert cache.type_.type_name() == 'struct kmem_cache *'
+    assert p.type_.type_name() == 'void *'
+    hardened_ptr = p + cache.offset.value_()
+
+    #
+    # We basically do what `freelist_dereference()` and
+    # `freelist_ptr()` do in the kernel source:
+    #
+    # ptr <- (void *)*(unsigned long *)(hardened_ptr)
+    #
+    intermediate_ulong = drgn.Object(sdb.get_prog(),
+                                     type='unsigned long',
+                                     address=hardened_ptr.value_())
+    ptr = drgn.Object(sdb.get_prog(),
+                      type='void *',
+                      value=intermediate_ulong.value_())
+
+    #
+    # ptr_addr <- (unsigned long)hardened_ptr
+    #
+    ptr_addr = drgn.Object(sdb.get_prog(),
+                           type='unsigned long',
+                           value=hardened_ptr.value_())
+
+    #
+    # return (void *)((unsigned long)ptr ^ cache->random ^ ptr_addr)
+    #
+    ptr_as_ulong = drgn.Object(sdb.get_prog(),
+                               type='unsigned long',
+                               value=ptr.value_())
+    clean_ptr_val = ptr_as_ulong.value_()
+    clean_ptr_val ^= cache.random.value_()
+    clean_ptr_val ^= ptr_addr.value_()
+    return drgn.Object(sdb.get_prog(), type='void *', value=clean_ptr_val)
+
+
+def cache_get_free_set(cache: drgn.Object, page: drgn.Object) -> Set[int]:
+    """
+    Return a set of addresses that point to all the freed
+    objects in this slab.
+    """
+    free_objects = []
+    p = page.freelist
+    while p != sdb.get_typed_null(p.type_):
+        free_objects.append(p.value_())
+        p = cache_get_free_pointer(cache, p)
+    return set(free_objects)
+
+
+def cache_gather_children(cache: drgn.Object) -> List[drgn.Object]:
+    children = []
+    for child in list_for_each_entry("struct kmem_cache",
+                                     cache.memcg_params.children.address_of_(),
+                                     "memcg_params.children_node"):
+        children.append(child)
+    return children
```

Comment:
same thing here, could this return an `Iterable` rather than a `List`, and be a generator?

Reply:
Changed to a Set[int] as discussed.